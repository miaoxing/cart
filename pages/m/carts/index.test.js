import Index from './index';
import {render, waitFor} from '@testing-library/react';
import $, {Ret} from 'miaoxing';
import {createPromise, bootstrap, setUrl, resetUrl} from '@mxjs/test';
import * as TaroTest from '@tarojs/taro';
import {createProduct} from '@miaoxing/product/test-utils';
import pretty from 'pretty';

bootstrap();
let didShow;
// eslint-disable-next-line no-import-assign
TaroTest.useDidShow = (fn) => {
  didShow = fn;
};

describe('Index', () => {
  beforeEach(() => {
    setUrl('/carts');
  });

  afterEach(() => {
    resetUrl();
  });

  test('empty', async () => {
    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: [],
        }),
      }));

    const {container, getByText} = render(<Index/>);
    didShow();

    await waitFor(() => {
      expect(getByText('购物车空空如也')).not.toBeNull();
    });

    expect(container).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });

  test('list', async () => {
    const product = createProduct();

    const promise = createPromise();

    $.http = jest.fn()
      .mockImplementationOnce(() => promise.resolve({
        ret: Ret.suc({
          data: [
            {
              id: 1,
              productId: product.id,
              skuId: product.skus[0].id,
              quantity: 9,
              changedPrice: null,
              addedPrice: '9',
              configs: {},
              updatedAt: '2021-06-28 16:45:59',
              sku: product.skus[0],
              product: product,
              createOrder: Ret.suc(),
            },
            {
              id: 2,
              productId: product.id,
              skuId: product.skus[1].id,
              quantity: 9,
              changedPrice: null,
              addedPrice: '9',
              configs: {},
              updatedAt: '2021-06-28 16:45:59',
              sku: product.skus[1],
              product: product,
              createOrder: Ret.err('该商品已失效'),
            },
            {
              id: 3,
              productId: product.id,
              skuId: product.skus[2].id,
              quantity: 9,
              changedPrice: null,
              addedPrice: '9',
              configs: {},
              updatedAt: '2021-06-28 16:45:59',
              sku: product.skus[2],
              product: product,
              createOrder: Ret.err('该规格已失效'),
            },
          ],
          selected: [1, 3],
        }),
      }));

    const {container, getByText} = render(<Index/>);
    didShow();

    await waitFor(() => {
      expect(getByText('该商品已失效')).not.toBeNull();
    });

    const html = pretty(container.innerHTML.replace(/id="swipeAction(.+?)"/g, 'id="REPLACED"'));

    expect(html).toMatchSnapshot();
    expect($.http).toMatchSnapshot();
  });
});
