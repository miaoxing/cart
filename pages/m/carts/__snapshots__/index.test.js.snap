// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index empty 1`] = `
<div>
  <taro-view-core
    class="mx-page css-slwbq5-Component"
    style="--mx-colors-gray-50: #fafafa; --mx-colors-gray-100: #f4f4f5; --mx-colors-gray-200: #e4e4e7; --mx-colors-gray-300: #d4d4d8; --mx-colors-gray-400: #a1a1aa; --mx-colors-gray-500: #71717a; --mx-colors-gray-600: #52525b; --mx-colors-gray-700: #3f3f46; --mx-colors-gray-800: #27272a; --mx-colors-gray-900: #18181b; --mx-space-0: 0; --mx-space-1: 4; --mx-space-2: 8; --mx-space-3: 12; --mx-space-4: 16; --mx-space-5: 20; --mx-space-6: 24; --mx-space-7: 28; --mx-space-8: 32; --mx-space-9: 36; --mx-space-10: 40; --mx-space-11: 44; --mx-space-12: 48; --mx-space-14: 56; --mx-space-16: 64; --mx-space-20: 80; --mx-space-24: 96; --mx-space-28: 112; --mx-space-32: 128; --mx-space-36: 144; --mx-space-40: 160; --mx-space-44: 176; --mx-space-48: 192; --mx-space-52: 208; --mx-space-56: 224; --mx-space-60: 240; --mx-space-64: 256; --mx-space-72: 288; --mx-space-80: 320; --mx-space-96: 384; --mx-fontSizes-xs: 12; --mx-fontSizes-sm: 14; --mx-fontSizes-base: 16; --mx-fontSizes-lg: 18; --mx-fontSizes-xl: 20; --mx-fontSizes-2xl: 24; --mx-fontSizes-3xl: 30; --mx-fontSizes-4xl: 36; --mx-fontSizes-5xl: 48; --mx-fontSizes-6xl: 60; --mx-fontSizes-7xl: 72; --mx-fontSizes-8xl: 96; --mx-fontSizes-9xl: 128;"
  >
    <taro-block-core>
      <taro-checkbox-group-core />
      <taro-view-core
        style="justify-content: center; align-items: center; color: rgb(113, 113, 122); flex-direction: column; display: flex;"
      >
        <taro-view-core
          class="mx-empty-image"
          style="height: 5.1rem;"
        />
        <taro-view-core>
          购物车空空如也
        </taro-view-core>
        <taro-view-core>
          <taro-navigator-core>
            <taro-button-core
              class="mx-btn mx-btn-primary"
            >
              去逛逛
            </taro-button-core>
          </taro-navigator-core>
        </taro-view-core>
      </taro-view-core>
      <taro-block-core />
    </taro-block-core>
    <taro-view-core
      class="mx-footer-bar"
      style="align-items: center; justify-content: space-between; flex-direction: row; display: flex;"
    >
      <taro-checkbox-group-core>
        <taro-label-core
          style="align-items: center; color: rgb(113, 113, 122); line-height: 1; flex-direction: row; display: flex;"
        >
          <taro-checkbox-core
            checked="true"
          />
          全选
        </taro-label-core>
      </taro-checkbox-group-core>
      <taro-view-core
        style="align-items: center; flex-direction: row; display: flex;"
      >
        <taro-view-core
          style="color: rgb(113, 113, 122);"
        >
          合计：
          <taro-text-core
            primary500="true"
          >
            <taro-text-core>
              ￥
            </taro-text-core>
            0
          </taro-text-core>
        </taro-view-core>
        <taro-view-core
          class="mx-btn-theme mx-btn-theme-pill"
        >
          <taro-button-core
            class="mx-btn mx-btn-primary"
          >
            结算（
            0
            ）
          </taro-button-core>
        </taro-view-core>
      </taro-view-core>
    </taro-view-core>
  </taro-view-core>
</div>
`;

exports[`Index empty 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/carts",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;

exports[`Index list 1`] = `
"<taro-view-core class=\\"mx-page css-slwbq5-Component\\" style=\\"--mx-colors-gray-50: #fafafa; --mx-colors-gray-100: #f4f4f5; --mx-colors-gray-200: #e4e4e7; --mx-colors-gray-300: #d4d4d8; --mx-colors-gray-400: #a1a1aa; --mx-colors-gray-500: #71717a; --mx-colors-gray-600: #52525b; --mx-colors-gray-700: #3f3f46; --mx-colors-gray-800: #27272a; --mx-colors-gray-900: #18181b; --mx-space-0: 0; --mx-space-1: 4; --mx-space-2: 8; --mx-space-3: 12; --mx-space-4: 16; --mx-space-5: 20; --mx-space-6: 24; --mx-space-7: 28; --mx-space-8: 32; --mx-space-9: 36; --mx-space-10: 40; --mx-space-11: 44; --mx-space-12: 48; --mx-space-14: 56; --mx-space-16: 64; --mx-space-20: 80; --mx-space-24: 96; --mx-space-28: 112; --mx-space-32: 128; --mx-space-36: 144; --mx-space-40: 160; --mx-space-44: 176; --mx-space-48: 192; --mx-space-52: 208; --mx-space-56: 224; --mx-space-60: 240; --mx-space-64: 256; --mx-space-72: 288; --mx-space-80: 320; --mx-space-96: 384; --mx-fontSizes-xs: 12; --mx-fontSizes-sm: 14; --mx-fontSizes-base: 16; --mx-fontSizes-lg: 18; --mx-fontSizes-xl: 20; --mx-fontSizes-2xl: 24; --mx-fontSizes-3xl: 30; --mx-fontSizes-4xl: 36; --mx-fontSizes-5xl: 48; --mx-fontSizes-6xl: 60; --mx-fontSizes-7xl: 72; --mx-fontSizes-8xl: 96; --mx-fontSizes-9xl: 128;\\">
  <taro-block-core>
    <taro-checkbox-group-core>
      <taro-view-core class=\\"mx-card\\">
        <taro-view-core class=\\"mx-list\\">
          <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action\\" style=\\"width: 0px;\\">
            <taro-movable-area-core class=\\"at-swipe-action__area\\" style=\\"width: 75px; transform: translate(-75px, 0);\\">
              <taro-movable-view-core class=\\"at-swipe-action__content\\" style=\\"width: 0px; left: 75px;\\">
                <taro-view-core class=\\"mx-list-item\\" style=\\"width: 100%;\\">
                  <taro-view-core class=\\"mx-list-col\\" style=\\"flex: 0 0px;\\">
                    <taro-view-core style=\\"align-items: center; flex-direction: row; display: flex;\\">
                      <taro-checkbox-core checked=\\"true\\"></taro-checkbox-core>
                      <taro-navigator-core style=\\"border-radius: 8rpx; overflow: hidden;\\">
                        <taro-image-core></taro-image-core>
                      </taro-navigator-core>
                    </taro-view-core>
                  </taro-view-core>
                  <taro-view-core class=\\"mx-list-col\\" style=\\"justify-content: space-between; flex-direction: column; display: flex;\\">
                    <taro-navigator-core truncate2=\\"true\\">alias commodi ipsam</taro-navigator-core>
                    <taro-block-core>
                      <taro-view-core>
                        <taro-view-core style=\\"display: inline-block; color: rgb(113, 113, 122); background-color: rgb(244, 244, 245); border-radius: 4rpx;\\">S；红色<taro-text-core class=\\"mxi mxi-chevron-down\\"></taro-text-core>
                        </taro-view-core>
                      </taro-view-core>
                      <taro-view-core style=\\"align-items: center; justify-content: space-between; flex-direction: row; display: flex;\\">
                        <taro-text-core primary500=\\"true\\">
                          <taro-text-core>￥</taro-text-core>9
                        </taro-text-core>
                        <taro-view-core>
                          <taro-view-core class=\\"mx-stepper\\">
                            <taro-button-core class=\\"mx-stepper-minus\\"></taro-button-core>
                            <taro-input-core class=\\"mx-stepper-input\\"></taro-input-core>
                            <taro-button-core class=\\"mx-stepper-plus\\" disabled=\\"true\\"></taro-button-core>
                          </taro-view-core>
                        </taro-view-core>
                      </taro-view-core>
                    </taro-block-core>
                  </taro-view-core>
                </taro-view-core>
                <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action__options\\" style=\\"transform: translate(75px, 0); opacity: 1;\\">
                  <taro-view-core class=\\"at-swipe-action__option\\">
                    <taro-text-core class=\\"option__text\\">移除</taro-text-core>
                  </taro-view-core>
                </taro-view-core>
              </taro-movable-view-core>
            </taro-movable-area-core>
          </taro-view-core>
        </taro-view-core>
      </taro-view-core>
      <taro-view-core class=\\"mx-card\\">
        <taro-view-core class=\\"mx-list\\">
          <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action\\" style=\\"width: 0px;\\">
            <taro-movable-area-core class=\\"at-swipe-action__area\\" style=\\"width: 75px; transform: translate(-75px, 0);\\">
              <taro-movable-view-core class=\\"at-swipe-action__content\\" style=\\"width: 0px; left: 75px;\\">
                <taro-view-core class=\\"mx-list-item\\" style=\\"width: 100%;\\">
                  <taro-view-core class=\\"mx-list-col\\" style=\\"flex: 0 0px;\\">
                    <taro-view-core style=\\"align-items: center; flex-direction: row; display: flex;\\">
                      <taro-checkbox-core disabled=\\"true\\"></taro-checkbox-core>
                      <taro-navigator-core style=\\"border-radius: 8rpx; overflow: hidden;\\">
                        <taro-image-core></taro-image-core>
                      </taro-navigator-core>
                    </taro-view-core>
                  </taro-view-core>
                  <taro-view-core class=\\"mx-list-col\\" style=\\"justify-content: space-between; flex-direction: column; display: flex;\\">
                    <taro-navigator-core truncate2=\\"true\\">alias commodi ipsam</taro-navigator-core>
                    <taro-block-core>
                      <taro-view-core>
                        <taro-view-core style=\\"display: inline-block; color: rgb(113, 113, 122); background-color: rgb(244, 244, 245); border-radius: 4rpx;\\">S；蓝色<taro-text-core class=\\"mxi mxi-chevron-down\\"></taro-text-core>
                        </taro-view-core>
                      </taro-view-core>
                      <taro-view-core style=\\"align-items: center; justify-content: space-between; flex-direction: row; display: flex;\\">
                        <taro-text-core primary500=\\"true\\">
                          <taro-text-core>￥</taro-text-core>11
                        </taro-text-core>
                        <taro-view-core>
                          <taro-view-core class=\\"mx-stepper\\">
                            <taro-button-core class=\\"mx-stepper-minus\\"></taro-button-core>
                            <taro-input-core class=\\"mx-stepper-input\\"></taro-input-core>
                            <taro-button-core class=\\"mx-stepper-plus\\" disabled=\\"true\\"></taro-button-core>
                          </taro-view-core>
                        </taro-view-core>
                      </taro-view-core>
                    </taro-block-core>
                    <taro-view-core style=\\"color: rgb(255, 107, 107);\\">该商品已失效</taro-view-core>
                  </taro-view-core>
                </taro-view-core>
                <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action__options\\" style=\\"transform: translate(75px, 0); opacity: 1;\\">
                  <taro-view-core class=\\"at-swipe-action__option\\">
                    <taro-text-core class=\\"option__text\\">移除</taro-text-core>
                  </taro-view-core>
                </taro-view-core>
              </taro-movable-view-core>
            </taro-movable-area-core>
          </taro-view-core>
        </taro-view-core>
      </taro-view-core>
      <taro-view-core class=\\"mx-card\\">
        <taro-view-core class=\\"mx-list\\">
          <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action\\" style=\\"width: 0px;\\">
            <taro-movable-area-core class=\\"at-swipe-action__area\\" style=\\"width: 75px; transform: translate(-75px, 0);\\">
              <taro-movable-view-core class=\\"at-swipe-action__content\\" style=\\"width: 0px; left: 75px;\\">
                <taro-view-core class=\\"mx-list-item\\" style=\\"width: 100%;\\">
                  <taro-view-core class=\\"mx-list-col\\" style=\\"flex: 0 0px;\\">
                    <taro-view-core style=\\"align-items: center; flex-direction: row; display: flex;\\">
                      <taro-checkbox-core disabled=\\"true\\"></taro-checkbox-core>
                      <taro-navigator-core style=\\"border-radius: 8rpx; overflow: hidden;\\">
                        <taro-image-core></taro-image-core>
                      </taro-navigator-core>
                    </taro-view-core>
                  </taro-view-core>
                  <taro-view-core class=\\"mx-list-col\\" style=\\"justify-content: space-between; flex-direction: column; display: flex;\\">
                    <taro-navigator-core truncate2=\\"true\\">alias commodi ipsam</taro-navigator-core>
                    <taro-block-core>
                      <taro-view-core>
                        <taro-view-core style=\\"display: inline-block; color: rgb(113, 113, 122); background-color: rgb(244, 244, 245); border-radius: 4rpx;\\">M；红色<taro-text-core class=\\"mxi mxi-chevron-down\\"></taro-text-core>
                        </taro-view-core>
                      </taro-view-core>
                      <taro-view-core style=\\"align-items: center; justify-content: space-between; flex-direction: row; display: flex;\\">
                        <taro-text-core primary500=\\"true\\">
                          <taro-text-core>￥</taro-text-core>12
                        </taro-text-core>
                        <taro-view-core>
                          <taro-view-core class=\\"mx-stepper\\">
                            <taro-button-core class=\\"mx-stepper-minus\\"></taro-button-core>
                            <taro-input-core class=\\"mx-stepper-input\\"></taro-input-core>
                            <taro-button-core class=\\"mx-stepper-plus\\" disabled=\\"true\\"></taro-button-core>
                          </taro-view-core>
                        </taro-view-core>
                      </taro-view-core>
                    </taro-block-core>
                    <taro-view-core style=\\"color: rgb(255, 107, 107);\\">该规格已失效</taro-view-core>
                  </taro-view-core>
                </taro-view-core>
                <taro-view-core id=\\"REPLACED\\" class=\\"at-swipe-action__options\\" style=\\"transform: translate(75px, 0); opacity: 1;\\">
                  <taro-view-core class=\\"at-swipe-action__option\\">
                    <taro-text-core class=\\"option__text\\">移除</taro-text-core>
                  </taro-view-core>
                </taro-view-core>
              </taro-movable-view-core>
            </taro-movable-area-core>
          </taro-view-core>
        </taro-view-core>
      </taro-view-core>
    </taro-checkbox-group-core>
    <taro-block-core></taro-block-core>
  </taro-block-core>
  <taro-view-core class=\\"mx-footer-bar\\" style=\\"align-items: center; justify-content: space-between; flex-direction: row; display: flex;\\">
    <taro-checkbox-group-core>
      <taro-label-core style=\\"align-items: center; color: rgb(113, 113, 122); line-height: 1; flex-direction: row; display: flex;\\">
        <taro-checkbox-core checked=\\"true\\"></taro-checkbox-core>全选
      </taro-label-core>
    </taro-checkbox-group-core>
    <taro-view-core style=\\"align-items: center; flex-direction: row; display: flex;\\">
      <taro-view-core style=\\"color: rgb(113, 113, 122);\\">合计：<taro-text-core primary500=\\"true\\">
          <taro-text-core>￥</taro-text-core>81
        </taro-text-core>
      </taro-view-core>
      <taro-view-core class=\\"mx-btn-theme mx-btn-theme-pill\\">
        <taro-button-core class=\\"mx-btn mx-btn-primary\\">结算（1）</taro-button-core>
      </taro-view-core>
    </taro-view-core>
  </taro-view-core>
</taro-view-core>"
`;

exports[`Index list 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "url": "/api/carts",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "reject": [Function],
        "resolve": [Function],
      },
    },
  ],
}
`;
